### ISSUE TITLE
Node Categories

### ISSUE THREAD

**Issue by Muzz** [HungryProton/concept_graph#54]
So far the nodes are relatively organized but there is inconsistency. 

I would like to see one core idea driving what category handles what. 
For example, there is the input category, which only has generic input, and the curve input is in curves. 

There is two competing ideas there, grouping by data type, or grouping by purpose. What do you see as being a better idea. 

Thinking about it, both are really fine solutions, but I would like to open this discussion.

**Comment by HungryProton** [comment]
I'm bad at organizing and naming things in general. I tried to balance how many node there is under each categories but it just makes things confusing.

But then, Blender does it too. In the material editor, you have data type categories like Color or Vector, but also purpose categories like Converter or Input. 
![image](https://user-images.githubusercontent.com/52043844/82382603-893df900-9a2c-11ea-80a6-c889cced1f73.png)

Grouping everything under the data type also means that most nodes will fall under the "Nodes" category (like everything that only affect transforms) while grouping everything based on their purpose means we will get a lot of categories with a single or very few nodes in it.

Another possible approach would be grouping by data type in top level categories and grouping by purpose under each. Not sure if there would be conflicts though, nodes that would fall under two different types at once (I can't think of an example right now so maybe it's fine?)

**Comment by Muzz** [comment]
Sorry for the late reply. 

I've been thinking about this more. 

I feel like a workflow oriented perspective on the grouping might make the most sense. 

When I'm using nodes I think about what am I trying to do and it could be any one of these.

- Get data into the node system
- Get data out of the node system
- Perform mathematics 
- Transform the data i have in some way
- Generate one form of data from another form. 

So if you take that logic, then you have,
- Input 
- Output
- Maths
- Modifiers
- Generators

I think this is already close to what's there, but there isn't a commitment to using this system. As there is mesh and curve categories. 

Not sure how practical it may be, but what may actually work better is if you have filters as well for type. So all the mesh and curve modifiers would live in the modifiers section, but at the top of the node dropdown would be toggles that filter by type. 

Another quality of life thing could be to colour code the labels in the dropdown, according to that type. 

I can do some mockups for how i see this working if you'd like.

**Comment by HungryProton** [comment]
That's close to what we have yes, I think it could work. Also we have 2D nodes now, do you think they should be mixed with the rest of the nodes or put in a completely separate category? 

I think the search bar should also be able to search categories too not only the nodes name, that should fix the filter issue and make the search behavior more consistent with the `Add child` panel for scene tree nodes. 

Coloring the label themselvs may or may not make them harder to read, but we could add a small icon next to each node with the proper color, again to make things more consistent with the rest of the editor. Let me know what you think.

**Comment by BryceBubbles** [comment]
**Categories**
This is a tricky one as both  "purpose" and "node type" categories are useful/necessary, and which comes first is likely down to personal preference. 

@HungryProton could we add some sort of "Category Order" setting that switches between "Node Type" and "Purpose" as top level? I'm thinking a toggle on the node panel itself, or perhaps on the settings panel you mentioned?

As for implementation, this could just reverse the order of the nodes category property?
Or perhaps we need to enforce the two levels by adding a "purpose" property and only allowing a single level on the category property?

**Search**
Because there's so many nodes, making search work well is going to be really important.

Being able to search by category would be great.
But I'd love to be able to search by **input and output type** and perhaps names as well?

**Colouring**
This is something I've been meaning to mention as well
Icons for consistency with Godot will be nice.

**Comment by HungryProton** [comment]
If the category nesting makes sense when reading it both ways, adding a toggle to invert them is easy, I'm afraid it will make tutorials harder to follow if people use different preset, but I guess this isn't a problem if the UI is clear enough.


So, taking in account everything that was said so far, **we would have something like this:**

+ Input
  - Curves
    + Curve Input
  - Meshes
  - ...
+ Output
  - Output 2D
  - Output 3D
  - ...
+ Inspector
+ Maths
+ Modifiers
  - Transforms
    - 2D
    - 3D
  - Meshes
    + Deform along curve
  - Curves
  - Heightmaps
+ Generators
  - Transforms
    + 2D
      - Create Point Grid 2D
    + 3D
      - Create Point Grid 3D
  - Meshes
    + Extrude Along Curve

**Which can also be read the other way around**, without affecting single level categories.

+ 2D
  - Transforms
    + Generators
    + Modifiers
+ 3D
  - Transforms
    + Generators
    + Modifiers
+ Meshes
  - Generators
    + Extrude Along Curve
  - Modifiers
    + Deform Along Curve
+ Maths
+ Vectors
+ Outputs
+ ...

 I think Vector can be a top level one, unless it makes more sense to go with : 

+ Generators
  - Vector
    + Make Vector3
  - Scalar
    + Break Vector3

Which feels a bit weird to me because `Break Vector` would fall under the scalar category (since it outputs 3 numbers), but makes it more consistent with the rest. And since search works now it shouldn't be hard to find.

(Also, should categories be plural or singular? I'm not a native english speaker, I don't know what feels more natural)

**Search**
I agree that the search needs to be more powerful, there's a lot more nodes coming and it probably needs its own issue to discuss this.

**Comment by BryceBubbles** [comment]
@HungryProton that all looks good to me.

**Plural vs Singular**
Don't worry, most native English speakers don't know either - lol I'm joking, but only sort of.

**Note that Blender uses singular in the image above.**
This is because using singular is much simpler.
Using plural adds complexity, with multiple variations (s, es, ies) and subject issues, where "Inspectors" should be "Inspector Properties" (like it was before I changed it to singular Inspector). 
Basically plurality adds visual and cognitive noise to lists.

However, "Generator" and "Modifier" feel a bit weird to me. Perhaps it's just that "Generators" and "Modifiers" are the common usage in software? Consistency is probably better though?

@Muzz what's you're thoughts?

**Comment by HungryProton** [comment]
I'm probably going to tackle this issue soon. The script files themselves all all mixed up in the nodes folder too so there's going to be a lot of renaming / moving files around. I guess I'm going to use the same grouping convention for the file system too.

**Comment by HungryProton** [comment]
I tried to summarize this discussion in this page, let me know if there's anything you'd like to add or edit : https://github.com/HungryProton/concept_graph/wiki/Naming-Conventions

I'm going to rename and move everything, hopefully the source folder should be easier to navigate.

**Comment by BryceBubbles** [comment]
**Categories singular and plural**
I notice in that doc you've listed the categories with some plural and some singular.
I'm just mentioning it based on above comments, but I'm fine with it either way.

**Node Names**
"Purpose first, then data" - possibly add a couple of examples?

**Class name prefix for namespacing.**
It's usually `ConceptGraph` but there's also `ConceptNode`, so is the prefix actually `Concept`?
- `ConceptGraph`
- `ConceptNode`
- `ConceptNodeLibrary`
- `ConceptGraphNodePool `- could this be `ConceptNodePool`?
- `ConceptGraphDataType` - could this be `ConceptDataType`
- `PoolVectorUtil` currently has no prefix

**Underscores for inheritance?**
For the current noise changes I used `ConceptGraphNoise` and `ConceptGraphNoise_Simplex` but perhaps it should be without `ConceptGraphNoiseSimplex`? Or `ConceptNoiseSimplex`?

**Class file names match class name?**
Per the Godot styleguide:
> Use snake_case for file names. For named classes, convert the PascalCase class name to snake_case.

Currently many, but not all file names drop the prefix, which does make them shorter.
But perhaps with a shorter prefix of "Concept", all files should then match the class name exactly?

**Comment by HungryProton** [comment]
**Categories singular and plural**
I went with plural for now since other software does it too, I'm not really sure how to pluralize some of them (like Flow Control, maybe it should be a completely different name too).

**Node Names**
I'm still not sure about the node name convention so this part is rather empty. There's currently a few different patterns, like `Transform (Random)` with the target first and the type of operation between parenthesis, and other that starts with a verb like `Make Multimesh`. Maybe the naming should depends on the category too? I haven't really thought about it yet.

**Class name prefix**
I wish GDscript had namespacing.
I started to prefix anything related to the base class by `ConceptGraph`, and everything related to the GraphNode `ConceptNode`, then I couldn't draw the line between the two for some of the classes so I guess simply using `Concept` for everything would work and make every class name shorter too.

**Inheritance**
I'm not really a fan of underscores in pascal case names and the official GDscript style guide forbids it anyway.

**Class file names**
If it's in the style guide it should be done this way then, one less question to worry about :)

**Comment by BryceBubbles** [comment]
**Plural**
That's the thing with plural, it requires more thought :) 
Technically the plural is `Flow Controls`, but `Flow Control` feels better to me, even alongside other plural categories. It's tricky.
Another name could be `Logic Gates` or just `Gates`? But `Flow Control` probably makes more sense?

**Node Names**
I think go for what creates the most visual consistency in the nodes menu.
For Generators and Modifiers verb first often works well - Create, Break, Exclude, Extrude, Merge, Set
_To be consistent `Make` should be `Create`._

`Transform (Random)` - technically "Transform" is the target, but "Transform" is also an operation on a node. It's more clear with "Position" (Translate?), "Rotate" and "Scale" that they're operations on a target.

But then we have `Transform (Break)` where break is the operation and transform is the target, so that should be `Break Transform` perhaps? But then the way the transform nodes are currently named is quite visually consistent, which I think is more important.

`Transform: Random` might be nicer though?

**Comment by HungryProton** [comment]
Three days without internet so I couldn't answer sooner.

+ The `Flow Control` category groups all the nodes that change the execution path (there's only conditionals in there for now, not sure how loops would work with the current system though). Logic Gates sounds more like the logic AND, OR and so on.

+ **Node names** rules sounds good to me, to summarize:
  - Verb first, then the Target (when possible, mostly with generators and modifiers)
  - Input first then the Output (converters and everything else)
+ Break Transform isn't really like the others, it just expose some vectors instead of editing an existing transform. It's actually the Transform nodes that don't follow any set of rules for now. Transform being both the verb and the target, I guess it's an exception.
+ `Transform: Random` or `Transform - Random` are both fine to me

**Comment by HungryProton** [comment]
I'm almost done with the new grouping options.

![image](https://user-images.githubusercontent.com/52043844/84565700-39bcc580-ad6b-11ea-80e4-b7ad2a0d1376.png)

+ Added an icon next to the nodes so they're easier to spot, especially now with all the new nested categories.
+ Checkbox on top to toggle between type and purpose grouping
  -  There's also an option in the Project Settings to set the default behavior

Let me know if there's anything else that could be improved.
